namui/src/namui/fs/bundle/read_dir/bundle_dir_reader.rs

```diff
@@ -0,0 +1,168 @@
+        self.path_dirent_list_map
+            .get(path)
+            .map(|dirent_list| dirent_list.clone())
+    }
+
+    async fn load(self: Arc<Self>) -> Result<(), LoadBundleMetadataError> {
+        self.set_load_state(BundleDirReaderLoadState::Loading);
+        let load_result = match load_bundle_metadata().await {
+            Ok(bundle_metadata) => {
+                let path_dirent_list_map = make_path_dirent_list_map(&bundle_metadata);
+                self.path_dirent_list_map.clear();
+                for (path, dirent_list) in path_dirent_list_map {
+                    self.path_dirent_list_map.insert(path, dirent_list);
+                }
+                self.set_load_state(BundleDirReaderLoadState::Loaded);
+                Ok(())
+            }
+            Err(error) => {
+                self.set_load_state(BundleDirReaderLoadState::Unset);
+                Err(error)
+            }
+        };
+        self.wake_wakers();
+        load_result
+    }
+
+    fn wake_wakers(&self) {
+        let mut load_wakers = self.load_wakers.write();
+        for waker in load_wakers.iter() {
+            waker.clone().wake();
+        }
+        load_wakers.clear();
+    }
+
+    fn push_waker(&self, waker: Waker) {
+        let mut load_wakers = self.load_wakers.write();
+        load_wakers.push(waker);
+    }
+}
+impl BundleDirReaderRead for Arc<BundleDirReader> {
+    fn read(&self, path_like: impl PathLike) -> BundleDirReaderReadTask {
+        let path = path_like.path();
+        BundleDirReaderReadTask {
+            path,
+            bundle_dir_reader: self.clone(),
+            load_future: None,
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+enum BundleDirReaderLoadState {
+    Unset,
+    Loading,
+    Loaded,
+}
+
+#[derive(Debug)]
+pub enum BundleDirReaderError {
+    NetworkError(String),
+    ParseError(String),
+    DirNotExist,
+    MetadataFileNotFound(String),
+    Other(String),
+}
+impl From<LoadBundleMetadataError> for BundleDirReaderError {
+    fn from(error: LoadBundleMetadataError) -> Self {
+        match error {
+            LoadBundleMetadataError::NetworkError(message) => {
+                BundleDirReaderError::NetworkError(message)
+            }
+            LoadBundleMetadataError::ParseError(message) => {
+                BundleDirReaderError::ParseError(message)
+            }
+            LoadBundleMetadataError::FileNotFound(message) => {
+                BundleDirReaderError::MetadataFileNotFound(message)
+            }
+            LoadBundleMetadataError::Other(message) => BundleDirReaderError::Other(message),
+        }
+    }
+}
+
+pub trait BundleDirReaderRead {
+    fn read(&self, path_like: impl PathLike) -> BundleDirReaderReadTask;
+}
+
+pub struct BundleDirReaderReadTask {
+    path: PathBuf,
+    bundle_dir_reader: Arc<BundleDirReader>,
+    load_future: Option<LocalBoxFuture<'static, Result<(), LoadBundleMetadataError>>>,
```

Comment:
Ah. I got it. As you said, it looks complicated.
In namui, I used to load metadata about namui in namui::init or namui::start. for example, the font.
I beilieve you can load bundle metadata at there, and I think it's good idea because bundle metadata isn't changed during runtime, isn't it?

Reply:
Oh yes, we already had Init procedure. And bundle metadata is better suited to be there.
I'll rework it.